src/fair/allocation.py

```diff
@@ -150,10 +150,12 @@ def find_shortest_path(G, start, end):
 def add_agent_to_exchange_graph(G, X, items, agents, agent_picked):
     G.add_node("s")
     bundle = get_bundle_from_allocation_matrix(X, items, agent_picked)
-    for i in range(len(items)):
+    agent = agents[agent_picked]
+    for i in agent.get_desired_items_indexes(items):
         g = items[i]
-        if agents[agent_picked].marginal_contribution(bundle, g) == 1:
-            G.add_edge("s", i)
+        if g not in bundle:
+            if agents[agent_picked].marginal_contribution(bundle, g) == 1:
```

Comment:
Prefer `if A and B` to nested if statements

Reply:
Nevermind, Python lazily checks conditions, I'll use an AND statement instead.